File: sdks/android/src/main/kotlin/dev/nucleus/models/NucleusSession.kt
package dev.nucleus.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Represents an active Nucleus session with access and refresh tokens.
 */
@Serializable
data class NucleusSession(
    val id: String,
    @SerialName("user_id")
    val userId: String,
    @SerialName("access_token")
    val accessToken: String,
    @SerialName("refresh_token")
    val refreshToken: String? = null,
    @SerialName("expires_at")
    val expiresAt: Long,
    @SerialName("created_at")
    val createdAt: Long? = null,
    @SerialName("last_active_at")
    val lastActiveAt: Long? = null,
    val status: String = "active",
) {
    /** Returns `true` when the access token has expired (with a 30-second safety margin). */
    val isExpired: Boolean
        get() = System.currentTimeMillis() / 1000 >= expiresAt - 30
}

File: sdks/swift/Sources/NucleusSwift/Models/NucleusSession.swift
import Foundation

/// Represents an active authentication session with access and refresh tokens.
public struct NucleusSession: Codable, Equatable, Sendable {
    public let accessToken: String
    public let refreshToken: String
    public let expiresAt: Date
    public let tokenType: String?

    /// Whether the access token has expired.
    public var isExpired: Bool {
        Date() >= expiresAt
    }

    /// Whether the access token will expire within 60 seconds.
    public var isExpiringSoon: Bool {
        Date().addingTimeInterval(60) >= expiresAt
    }

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresAt = "expires_at"
        case tokenType = "token_type"
    }
}
